### DISCUSSION TITLE
How to run pikaraoke as a Kiosk on FullPageOS (Oct 2024)

### DISCUSSION THREAD

**Discussion by Romeo1984** [vicwomg/pikaraoke#406] [Show and tell]
So, I've been running PiKaraoke since it was first posted. It has undergone a necessary evaluation that allowed for ease of supportability and portability across platforms. My particular use case was to have a Pi device that was a dedicated karaoke machine that was easy to use and share with friends who also liked my parties and wanted to have their own device for their own parties. The original pikaraoke version can be installed on a server-only OS that output video directly to the video port on my pi4 device. I also have 65,000 CDG files that I had to have because there were a lot of obscure songs that don't have a more modern MP4 version out there on YouTube. 

Because there are several people here who have the same use case as me, I decided to create this How-To that achieves both adaptability and functionality in line with what the code authors are trying to do here. I intend to care and feed this thread as long as I can keep up with code releases. 

**What this How-To is NOT:**
If you have a different requirement, then this is not a help board. Please seek your own solution. You can post your issue as long as you found a solution and you think others can benefit from it. As an example, WiFi configuration seems to be an on-going issue for some people. Being I use Ethernet, I a not interested in seeing issues about your WiFi configuration - Please use the "Issues" section.

Here we go!

**My Requirements:**
1. Pi 4
2. Argon Case and Remote
3. The device is hooked up to an HDMI port on my 65" TV full-time
4. Use the 3.5mm headphone jack - Connected to an EQ and mixer, then to a stereo system
5. Needs to play both MP4 and CDG files
6. No system boot-up messages
7. Party members can also access pikaraoke on my private network
8. The device connects to Ethernet
9. No Desktop Env

**Skills required:**
1. Know how to install an OS onto a pi device 
2. Have an understanding of Linux
3. Can use some SSH client
4. Can use Nano or VI to edit files

**My Bill of Materials:**
1. Pi4 4 GB RAM
2. Argon-1 Case
3. Argon SSD add-on (They also have an NVE version)
4. Argon Remote
5. 500 GB SSD Stick (because I have 300GB in karaoke files)

**Versions**
1. Platform == Raspberry Pi 4 Model B Rev 1.4
2. PiKaraoke == 1.5.1
3. OS == Debian 1:6.6.20-1+rpt1 (2024-03-07) (bookworm)
4.  chromium-browser == Chromium 122.0.6261.89 Built on Raspbian , running on Raspbian 12
5. ffmpeg == 5.1.6-0+deb12u1+rpt1
6. yt-dlp == 2024.10.07

**Steps:** 
1. Switch the Pi-4 boot source from SD to USB
7. Install the OS Image on the 500GB drive. FullPageOS Image from here: https://unofficialpi.org/Distros/FullPageOS/nightly/
(I used 2024-03-31_2024-03-15-fullpageos-bookworm-armhf-lite-0.14.0.zip)
8. Get the IP address and SSH of the device. I use my networking system to see what new IP address is on my network. 
9. Update the OS then reboot:
`sudo apt-get update && sudo apt-get upgrade -y`
`sudo reboot`
10. Install ffmpeg:
`sudo apt-get install ffmpeg -y`
11. Follow the pikaroke install instructions. Create a .venv directory in the homedir:
`python -m venv ~/.venv`
Activate your virtual environment
`source ~/.venv/bin/activate`
Install pikaraoke from PyPi
`pip install pikaraoke`
12. Create the systemd service:
`sudo nano /etc/systemd/system/pikaraoke.service`

[Unit]
Description=pikaraoke
After=multi-user.target

[Service]
WorkingDirectory=/home/pi/.venv/
Type=simple
ExecStart=/home/pi/.venv/bin/python /home/pi/.venv/bin/pikaraoke --headless  --download-path /home/pi/pikaraoke-songs --volume 0.75 --high-quality
User=root
Group=root
Environment="PATH=$PATH:/home/pi/.venv/bin:/home/pi/.venv/lib/python3.11/site-packages:/usr/bin:/usr/sbin"
Restart=always
RestartSec=3

[Install]
WantedBy=multi-user.target

13. Install the service:
`sudo systemctl enable pikaraoke.service`

**Troubleshooting tip:** If you have to make changes to the pikaraoke.service file, you have to reload the service in order for the changes to load. Use the following command:

`sudo systemctl --system daemon-reload`

14. Switch the default Kiosk page to pikaraoke splash page:
`sudo nano /boot/firmware/fullpageos.txt`

Replace the default page with this:
`http://localhost:5555/splash?confirm=false` 

15. Reboot and see if pikaraoke launches:
`sudo reboot`

At this stage, we still need to work on the audio and adjust a few things to get the full Kiosk experience. The splash page can be seen as very small depending on the size of the TV you are working on. My lab has a 19" that looks okay, but when I put it up on my 65" in the living room, the splash page looks tiny. We will fix this shortly. 

16. Get the audio working:
a. Set the Audio Out to 3.5 Headphone Jack
`sudo raspi-config`
`1 System Options -> S2 Audio -> 69 Headphones -> OK -> Finish` 
b. Setup ALSA and default volume - The system volume needs to be set or you may experience your speakers overdriving if you try and adjust the karaoke volume in the application. 
Check devices by running beacue this may move depending on which OS or pi device you are using:  
`aplay -l` 
Look to see what device and subdevices the bcm2835 Headphones jack is on:

card 0: vc4hdmi0 [vc4-hdmi-0], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: vc4hdmi1 [vc4-hdmi-1], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 2: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones] **<-- Card #2 Device #0**
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7

Edit the ALSA config to match the above settings: 
`sudo nano /usr/share/alsa/alsa.conf`

defaults.ctl.card 2
defaults.pcm.card 2
defaults.pcm.device 0
defaults.pcm.subdevice 0

Intsall the PulseAudio utilities so we can set the default volume (your level may be different. Start with 70%) 

`sudo apt install pulseaudio -y`
`sudo reboot`

Set the volume:
`amixer set Master 70%`
`sudo alsactl store`

**At this point, pikaraoke should work 100%. Try searching for a YouTube file to play.** 

17. Setup for more of a Kiosk experience:
a. Change the Loading Page (Here is my custom graphic):
Click to download and Transfer the "loading" graphic file via SCP to '/home/pi/loading.png' (The full graphic is obscured here):
![loading](https://github.com/user-attachments/assets/94713da7-2afd-4750-8a3c-9241667b08af)

`sudo nano /opt/custompios/scripts/run_onepageos`

feh --bg-scale /home/pi/loading.png
#feh --bg-center /opt/custompios/background.png

![363741102-f77131ab-808d-481d-9d33-e375cf452075](https://github.com/user-attachments/assets/4a7d4714-f759-4405-b713-93f01605cdf6)

Trouble-shooting Tip: If you turn on your pikaraoke device and this screen hangs (pikaraoke does not load) then it means there was an issue. using SSH query the service and logs to see what the error is:

`sudo service pikaraoke status`

`journalctl -u pikaraoke.service`

b. For big TVs you will need to increase the size of the pikaraoke splash page.  Mine is a 65" Samsung 4k, so, these settings worked for me. You might have to adjust the factor multiplier to fit your size of TV.

 `sudo nano ~/scripts/start_chromium_browser`

Add to the "flags=()" section:

--force-device-scale-factor=4.0 %U

c. Remove the Linux system messages when booting up:
`sudo nano /boot/firmware/cmdline.txt`

 Change:  consoleblank=1 

18. (OPTIONAL) Install the Samba share to transfer your awesome karaoke library to (Note that the User will depend on your OS install):

I like to organize my MP4s and CDGs in separate directories. Notice that the pikaraoke startup switch will send the downloaded MPs to '/home/pi/pikaraoke-songs/'. 

Create the directories and apply world permissions (this is for ease of functionality. You can tighten this down if you wish): 

`mkdir /home/pi/pikaraoke-songs/CDG`
`mkdir /home/pi/pikaraoke-songs/YT`
`sudo chmod 0777 /home/pi/pikaraoke-songs/CDG`
`sudo chmod 0777 /home/pi/pikaraoke-songs/YT`
`sudo chmod 0777 /home/pi/pikaraoke-songs/`

`sudo apt-get install samba samba-common-bin`
`sudo nano /etc/samba/smb.conf`

Add the following section:
	[karaoke]
	comment = Shared folder
	path = /home/pi/pikaraoke-songs/
	browseable = yes
	writeable = yes
	only guest = no
	force user = pi **<-- This will be the user you configured in the OS that you have been signing in as**
	create mask = 0777
	directory mask = 0777
	public = yes
	guest ok = yes
 
`sudo service smbd restart`
`sudo service nmbd restart`

On your library computer, browse to \\<ip>\karaoke.
Transfer your files using the share. Sign in as your pi username. In this case, it will be "pi". We will disable this service later. 

19. (OPTIONAL) Install ARGON utilities for the Remote and Fan:
`wget https://download.argon40.com/argon1.sh`
`sudo chmod 700 argon1.sh`
`sudo ./argon1.sh`
`sudo argon-config`

Set the remote and fan to the desired parameters.

20. Disable unnecessary services (some may not exist):

`sudo systemctl disable lighttpd.service`
`sudo systemctl disable bluetooth.service`

`sudo systemctl disable x11vnc.service`
(This is a work around that corrects a bug in the OS build)
`sudo mkdir -p /var/lib/userconf-pi`
`sudo touch /var/lib/userconf-pi/autologin`
`sudo systemctl set-default graphical.target`

`sudo systemctl disable triggerhappy.service`
`sudo systemctl disable printing.service`
`sudo systemctl disable splashscreen.service`

(After your file transfer is complete) 
`sudo systemctl disable smbd.service` 
`sudo systemctl disable nmbd.service`

21. **That's it! Reboot and see if everything works!**

![366640863-9b338e65-e08a-4689-97df-d4ca71d3db6c](https://github.com/user-attachments/assets/d0b9629b-6b75-4bcc-b42d-7862028880cf)

**Comment by vicwomg** [comment]
Wow thanks so much for this! I’ll look it over and add it to the wiki

**Comment by vicwomg** [comment]
added to wiki: https://github.com/vicwomg/pikaraoke/wiki/Setting-up-FullPageOS-with-pikaraoke-(Oct-2024)
